CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+    record $fields =>
+        s = 0
+        for field in $fields
+            s = align_up_to(s, align(direction, field))
+            s += size(direction, field)
+        align_up_to(s, align(direction, record $fields))
+    variant $cases =>
+        s = 0
+        for case in $cases
+            s = max(s, size(direction, record([i32, case])))
+        s
+    shared-handle => size(i32)
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        size(record([i32, i32, i32]))
+      else:
+        size(i32)
+
+def align(direction, type):
+  match type
+    f32 = 4
+    f64 = 8
+    s8 = 1
+    u8 = 1
+    s16 = 2
+    u16 = 2
+    s32 = 4
+    u32 = 4
+    s64 = 8
+    u64 = 8
+    char = 4
+    list $ty = 4
+    record $fields =
+        a = 1
+        for field in $fields
+            a = max(a, align(direction, field))
+        a
+    variant $cases =
+        a = 4
+        for case in $cases
+            a = max(align, align(case))
+        a
+    shared-handle => align(i32)
+    push-buffer | pull-buffer => align(i32)
+```
+
+With those definitions in mind the way to read types from memory is then
+relatively straightforward:
+
+```
+def read_from_memory(direction, type, ptr):
+  # Note that like `lift`, no validation happens here.
+  match interface_type
+    f32 => read_4_bytes(ptr),
+    f64 => read_8_bytes(ptr),
+    s8 => read_1_byte(ptr),
+    u8 => read_1_byte(ptr),
+    s16 => read_2_bytes(ptr),
+    u16 => read_2_bytes(ptr),
+    s32 => read_4_bytes(ptr),
+    u32 => read_4_bytes(ptr),
+    s64 => read_8_bytes(ptr),
+    u64 => read_8_bytes(ptr),
+    char => read_4_bytes(ptr),
+
+    list $ty => {
+      ptr = read_4_bytes(ptr)
+      len = read_4_bytes(ptr + 4)
+      lift(direction, type, [ptr, len])
+    }
+
+    record $fields => {
+      if is_bitflags($fields):
+        if $fields.len() > 32:
+          lift(direction, type, [read_8_bytes(ptr)])
+        else:
+          lift(direction, type, [read_4_bytes(ptr)])
+
+      else:
+        fields = []
+        for field in $fields:
+          ptr = align_up_to(ptr, align(direction, field))
+          fields.push(read_from_memory(direction, field, ptr))
+          ptr += size(direction, field)
+        fields as $language_struct
+    }
+
+    variant $cases => {
+      discrim = read_4_bytes(ptr)
+      case = $cases[discrim]
+      ptr = align_up_to(ptr + 4, align(case))
+      payload = read_from_memory(case, ptr)
+
+      (discrim, payload) as $language_enum
+    }
+
+    shared-handle => lift(direction, type, [read_4_bytes(ptr)])
+
+    push-buffer | pull-buffer => {
+        assert(direction == "export")
+        lift(direction, type, [read_4_bytes(ptr)])
+    }
+```
+
+## Lowering values
+
+Lowering is very similar to lifting, it's basically just the opposite in various
+directions.
+
+```
+def lower_all(direction, types, values):
+    values = values.iter()
+    types.map(|ty| lower(direction, ty, values))
+
+def lower(direction, type, value):
```

Comment:
... and similarly here I'll assume this would be "Serializing language values into the canonical ABI" and `serialize`?